shared/src/balancer/event_handler.rs

```diff
@@ -77,24 +83,51 @@ trait PoolDataFetching: Send + Sync {
 impl PoolDataFetching for Web3 {
     /// Could result in ethcontract::{NodeError, MethodError or ContractError}
     async fn get_pool_data(&self, pool_address: H160) -> Result<WeightedPoolData> {
+        let mut batch = CallBatch::new(self.transport());
         let pool_contract = BalancerV2WeightedPool::at(self, pool_address);
         // Need vault and pool_id before we can fetch tokens.
         let vault = BalancerV2Vault::deployed(&self).await?;
         let pool_id = H256::from(pool_contract.methods().get_pool_id().call().await?.0);
-        let tokens = vault
+
+        // token_data and weight calls can be batched
+        let token_data = vault
             .methods()
             .get_pool_tokens(Bytes(pool_id.0))
-            .call()
-            .await?
-            .0;
+            .batch_call(&mut batch);
+        let normalized_weights = pool_contract
+            .methods()
+            .get_normalized_weights()
+            .batch_call(&mut batch);
+        batch.execute_all(MAX_BATCH_SIZE).await;
+
+        let tokens = token_data.await?.0;
+        // This is already a batch call for a list of token decimals.
+        let mut batch = CallBatch::new(self.transport());
```

Comment:
Why is this a separate CallBatch and not reusing the old one? Also, from reading the PR description I was assuming we would use the existing `TokenInfoFetching` component here?

Reply:
We started with using token info fetching, but it returns a hashmap which (I assume) is not guaranteed to preserve order. Rather than having to shuffle everything around, I decided not to use it. If it does preserve order, then I would be happy to use it.

Also, I use a new batch because I wanted to make sure it was empty for the next call. If this is not necessary then I woul also be happy to change this.

Reply:
I see now why we need to use a new batch. It is because the token data is part of the first call (thus the first batch needs to be executed first).

I still don't understand though, why TokenInfoFetching is not working in this case though. You can look up the tokens in order from your original list, can't you?

Generally I think we should be mindful of adding extra roundtrips (with the token fetcher we'd likely have some cache amortisation).

Reply:
I wasn't sure if HashMaps were guaranteed to be order preserving so I avoided it. I also avoided it because, regardless of the potential for Cached Token Info fetching, the only way we were able to use it was to instantiate a new one on each call anyway. I can imagine a situation where we instantiate a cached token info fetching external to the `BalancerEventUpdater` and pass it along all the way through and down to our PoolDataFetching so that this stuff is indeed cached and we can benefit from it. 

I would be happy to try making this happen as long as we are sure that the HashMap returned is order preserving. 

Will open a PR and see how it looks.

Reply:
This is turning out to be a super ugly change handing down and passing around a bunch of `Option<&CachedPoolFetchers>` everywhere, expected lifetime parameters and makes the PoolRegistry even harder to test (with mocking the pool fetchers). Do you really want this?

Reply:
I would pass in an `Arc<dyn TokenInfoFetching>` like we do e.g. in the `HttpSolver`. This way there shouldn't be the need for lifetimes (also no need to leak the fact that you want to use the cached version) and it should also only need to live on `BalancerPoolFetcher` which I assume will be instantiated from main (so I don't think it would be in many places).

Reply:
> I would be happy to try making this happen as long as we are sure that the HashMap returned is order preserving.

HashMaps don't preserve order. However, since you still have the order of your tokens from the input, you can look up the decimals in order from that list.